token-swap/program/src/curve/calculator.rs

```diff
@@ -145,6 +148,16 @@ pub trait CurveCalculator: Debug + DynPack {
     fn allows_deposits(&self) -> bool {
         true
     }
+
+    /// Calculates the total value of the curve given the liquidity parameters.
+    ///
+    /// This is mostly useful for invariant testing the curves, to make sure
+    /// that value is not lost on any trade.
+    ///
+    /// The default implementation for this function gives the Uniswap invariant
+    fn total_value(&self, swap_token_a_amount: u128, swap_token_b_amount: u128) -> Option<U256> {
+        U256::from(swap_token_a_amount).checked_mul(U256::from(swap_token_b_amount))
```

Comment:
Looks like token amounts can only be as high as u64::MAX so I don't think u256 is needed here

Reply:
This brought up another little bug in `PreciseNumber` -- we needed a max allowable base of `u128::MAX * ONE`